### ISSUE TITLE
[PyTorch] Printing large tensors is slow

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#6863]
Printing large tensors is slow in master.

```python
>>> x = torch.randn(1000, 1000, 1000)
>>> %timeit repr(x)
1 loop, best of 3: 21.8 s per loop
```

This is important because we often deal with very large tensors.

Note that printing large tensors is much faster in NumPy (10,000x in this case):

```python
>>> %timeit repr(x.numpy())
100 loops, best of 3: 2.18 ms per loop
```

**Comment by gchanan** [comment]
One thing I wanted to look at: when printing floating point tensors, we take the ceiling of the entire tensor to see if we can represent it as an integer tensor; we don't just do that on the numbers we are actually going to display.  I wonder what NumPy does here.

**Comment by gchanan** [comment]
Closed via https://github.com/pytorch/pytorch/pull/6876.

**Comment by sapjunior** [comment]
Could youe please pre-compiled binary fixed for this issue? It's really annoying

**Comment by fmassa** [comment]
Can't you use the nightly binaries?

**Comment by vadimkantorov** [comment]
On a recent master printing large GPU tensors is extremely slow, despite that a very small portion actually should get formatted.

```shell
python3 -c 'import torch; a = torch.load("bug.pt7"); print(a.device, a.shape)'
# cuda:0 torch.Size([1, 1700, 34, 50])

# GPU printing
time python3 -c 'import torch; print(torch.load("bug.pt7"))' > log.txt
# real    0m41.764s
# user    0m39.752s
# sys     0m2.016s

# only a few bytes are written in fact
wc -c log.txt
# 3771 log.txt

# CPU printing
time python3 -c 'import torch; print(torch.zeros(*torch.load("bug.pt7").shape))' > log.txt
# real    0m1.938s
# user    0m1.776s
# sys     0m0.676s
```

Can we reopen this? @colesbury @gchanan

[bug.pt7.zip](https://github.com/pytorch/pytorch/files/2620201/bug.pt7.zip)

**Comment by fmassa** [comment]
The problem in here is that your first dimension is of size 1, [and this part of the code](https://github.com/pytorch/pytorch/blob/master/torch/_tensor_str.py#L222) only splits multi-dim tensor if the number of elements in the first dimension is large. We should probably change this

**Comment by fmassa** [comment]
I'm looking into this

**Comment by soumith** [comment]
closed via https://github.com/pytorch/pytorch/pull/14418
